### ISSUE TITLE
Not including constant?

### ISSUE THREAD

**Issue by david-waterworth** [catboost/catboost#622]
I'm finding that `CatBoostRegressor` isn't minimising the mean squared error for my problem, I can trivially improve the mse by adding / subtracting a constant.

i.e.

```
offsets = np.arange(-15,25,.1)
errors = []
for offset in offsets:
    errors.append(mean_squared_error(y, best_estimator.predict(X)+offset))

plt.plot(offsets,errors)

offsets[np.argmin(errors)]
```

results in a parabolic plot with a minimum error for offset != 0

Specifically I can minimise the mse by calculating the offset as the difference between means of the predicted and target arrays, i.e.

`offset = np.mean(y)-np.mean(best_estimator.predict(X))`

Both [Friedman et. al. ](https://arxiv.org/pdf/0811.1679) (9), (10) and mboost perform boosting by starting with a constant and adding base learners - are you doing this?

**Comment by david-waterworth** [comment]
Adding to this, this is the example from your quick start - https://tech.yandex.com/catboost/doc/dg/concepts/python-quickstart-docpage/

```
import numpy as np
from catboost import Pool, CatBoostRegressor
# initialize data
train_data = np.random.randint(0, 
                               100, 
                               size=(100, 10))
train_label = np.random.randint(0, 
                                1000, 
                                size=(100))
test_data = np.random.randint(0, 
                              100, 
                              size=(50, 10))
# initialize Pool
train_pool = Pool(train_data, 
                  train_label, 
                  cat_features=[0,2,5])
test_pool = Pool(test_data, 
                 cat_features=[0,2,5]) 

# specify the training parameters 
model = CatBoostRegressor(iterations=2, 
                          depth=2, 
                          learning_rate=1, 
                          loss_function='RMSE')
#train the model
model.fit(train_pool)
# make the prediction using the resulting model
preds = model.predict(test_pool)
print(preds)
```

The prediction is biased - I've not set random state so it'll vary but if the algorithm has found the true minimum RMSE this should be zero right?

`np.mean(model.predict(train_pool))-np.mean(train_label)`

I'm not sure that a greedy tree construction does guarantee to minimise the rmse - I'm interested in your thoughts. In most papers CART splits on sum of squared error - which seems to be derived from the mse of the left and right split weighted by the frequency of each node. This seems to create a more balanced tree at the cost of some bias in the error?

This code solves for the constant which actually gives the minimum mse across the training set. Can you please provide some feedback?

```
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

offsets = np.arange(-15,25,.1)
errors = []
for offset in offsets:
    errors.append(mean_squared_error(train_label, model.predict(train_pool)+offset))

plt.plot(offsets,errors)

offsets[np.argmin(errors)]
```

**Comment by david-waterworth** [comment]
Any update on this? It's starting to concern me as I'm occasionally seeing very large errors and it's holding up production deployment.

For example today I observed the results below - this is from a model with a very poor fit due to very poor quality data. Notice that the score returned by the estimator (257) is very different to the last iteration score (133) and the mean of the residuals is very large (160) which is 15-20% of the mean of the target. I understand catboost transforms the predictors so that the cost it's minimising isn't the same as the score calculated on the original data, but surely it should always produce errors with zero mean?

```
from catboost import CatBoostRegressor

cat_features = [X.columns.get_loc(c) for c in X.columns if c in X.select_dtypes(include=['object']).columns]
estimator = CatBoostRegressor(loss_function='RMSE', has_time=True, logging_level='Verbose', task_type='GPU', devices='0', metric_period=100)

estimator.fit(X, y, cat_features)
print(f'score: {estimator.score(X,y)}')

constant = np.mean(y - estimator.predict(X))
print(f'mean(error): {constant}')
```

> 0:	learn: 819.5919022	total: 4.42ms	remaining: 4.42s
> 100:	learn: 195.9018572	total: 387ms	remaining: 3.45s
> 200:	learn: 184.0837204	total: 716ms	remaining: 2.85s
> 300:	learn: 172.7874290	total: 1.06s	remaining: 2.47s
> 400:	learn: 164.9883932	total: 1.52s	remaining: 2.27s
> 500:	learn: 158.0301906	total: 2.02s	remaining: 2.01s
> 600:	learn: 155.4061680	total: 2.47s	remaining: 1.64s
> 700:	learn: 150.8476749	total: 2.95s	remaining: 1.26s
> 800:	learn: 144.0193483	total: 3.47s	remaining: 861ms
> 900:	learn: 138.8180088	total: 3.98s	remaining: 438ms
> 999:	learn: 133.3853036	total: 4.5s	remaining: 0us
> score: 257.006055776171
> mean(error): -160.28277774797706

**Comment by annaveronika** [comment]
Working on that right now. Thanks a lot for the suggestion!

**Comment by annaveronika** [comment]
It's implemented in catboost 0.17. Flag is called start_from_average and is set to True by default.
